File: app/static/trade-ledger.js
/* Confirmed fills and reviewed VND marks, not TradingView model positions. */
(() => {
  const panel = document.querySelector('#tradeLedgerPanel');
  let identity = null, version = 0, latest = null;
  const esc = v => String(v ?? '').replace(/[&<>"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
  const num = v => v == null ? '—' : Number(v).toLocaleString('vi-VN',{maximumFractionDigits:2});
  const labels = {deposit:'Nạp tiền',withdraw:'Rút tiền',buy:'Mua khớp',sell:'Bán khớp',dividend:'Cổ tức tiền ròng',expense:'Chi phí khác'};
  const input = (name,label,type='text',value='',extra='') => `<label>${label}<input name="${name}" type="${type}" value="${esc(value)}" ${type==='number'?'step="any" min="0"':''} ${extra}></label>`;
  const table = (headers,rows) => `<div class="tableWrap"><table><thead><tr>${headers.map(h=>`<th>${esc(h)}</th>`).join('')}</tr></thead><tbody>${rows.length?rows.map(r=>`<tr>${r.map(c=>`<td>${esc(c)}</td>`).join('')}</tr>`).join(''):`<tr><td colspan="${headers.length}">Chưa có dữ liệu</td></tr>`}</tbody></table></div>`;
  async function api(path='',data) {
    const r = await fetch('/api/admin/ledger'+path,data===undefined?{}:{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify(data)});
    const body = await r.json();
    if(!r.ok) throw new Error(Array.isArray(body.detail)?body.detail.map(e=>`${e.loc.join('.')}: ${e.msg}`).join('; '):body.detail || `HTTP ${r.status}`);
    return body;
  }
  function download(name,data) {
    const url=URL.createObjectURL(new Blob([JSON.stringify(data,null,2)],{type:'application/json'}));
    const a=document.createElement('a');a.href=url;a.download=name;a.click();setTimeout(()=>URL.revokeObjectURL(url),1000);
  }
  async function action(fn,success) {
    const status=panel.querySelector('[data-status]');
    const currentIdentity=identity;
    status.textContent='Đang xử lý…';
    try {await fn(); if(identity!==currentIdentity)return; status.textContent=success;await load();}
    catch(e){if(identity===currentIdentity)status.textContent=e.message;}
  }
  function mount() {
    const day=new Date().toLocaleDateString('en-CA',{timeZone:'Asia/Ho_Chi_Minh'});
    panel.innerHTML=`<p class="eyebrow">Sổ thực tế • VND • quản trị viên</p><h2>Sổ giao dịch + NAV + đối soát dữ liệu</h2>
      <p>Webhook không tự sinh giao dịch khớp. Nhập theo sao kê: nạp/rút tiền, mua/bán thực khớp, phí và thuế. Không hỗ trợ tiền mặt âm, bán khống hoặc margin. Thứ tự cùng ngày là thứ tự nhập.</p>
      <div class="fundActions">${input('valuation_date','Ngày định giá','date',day,'required')}<button type="button" data-reload>Xem / đối soát</button></div>
      <div data-summary></div>
      <details><summary>Ghi giao dịch / dòng tiền</summary><form data-entry><div class="fundForm">
        ${input('reference','Mã đối chiếu duy nhất (sao kê)','text','','required maxlength="100"')}${input('trade_date','Ngày giao dịch','date',day,'required')}
        <label>Loại<select name="kind">${Object.entries(labels).map(([k,v])=>`<option value="${k}">${v}</option>`).join('')}</select></label>
        ${input('ticker','Mã cổ phiếu')}
        <label>Nhóm<select name="sleeve"><option>OTHER</option><option>RF</option><option>EMA</option></select></label>
        ${input('sector','Ngành','text','unknown')}${input('quantity','Số lượng khớp','number','0')}${input('price','Giá khớp VND','number','0')}${input('fees','Tổng phí + thuế VND','number','0')}${input('amount','Số tiền nạp/rút/cổ tức/chi phí VND','number','0')}${input('signal_id','ID webhook liên kết (tùy chọn)','number')}${input('note','Ghi chú')}
      </div><p>Mua/bán: amount = 0. Dòng tiền: quantity, price, fees = 0. Cổ tức là số tiền sau thuế. Nhiều lần khớp có thể cùng ID webhook, nhưng mỗi lần phải có mã đối chiếu riêng.</p><button>Lưu vào sổ</button></form></details>
      <details><summary>Kiểm chứng giá cuối ngày</summary><form data-price><div class="fundForm">
        ${input('trade_date','Ngày giá','date',day,'required')}${input('ticker','Mã','text','','required')}${input('price','Giá đóng cửa VND (ví dụ 25000)','number','','required min="0.000001"')}${input('source','Nguồn giá (sao kê/nhà cung cấp)','text','','required minlength="3"')}${input('evidence','Tham chiếu kiểm chứng / ghi chú','text','','required minlength="3"')}
      </div><label><input name="confirmed" type="checkbox" required> Đã đối chiếu đúng mã, ngày, đơn vị VND và giá đóng cửa chưa điều chỉnh</label><p>Giá nhà cung cấp/webhook chưa được tự động chứng nhận. Ghi giá mới cùng mã/ngày sẽ thay bản đang dùng nhưng giữ mọi phiên bản trong lịch sử.</p><button>Lưu giá đã kiểm chứng</button></form></details>
      <details><summary>Chốt NAV từ sổ giao dịch</summary><form data-close><div class="fundForm">${input('trade_date','Ngày chốt NAV','date',day,'required')}${input('benchmark_name','Benchmark','text','VNINDEX (price)','required')}${input('benchmark_value','Mức benchmark đã đối chiếu','number')}</div>
      <p>NAV = tiền mặt kế toán + số lượng × giá cùng ngày đã kiểm chứng. Sổ ghi nhận tiền theo ngày giao dịch, không phải tiền có thể rút theo T+2. Chốt sẽ khóa toàn bộ giao dịch/giá đến ngày này và cập nhật Hiệu suất NAV + Rủi ro. Không ghi đè NAV có sẵn. Dòng tiền giữa hai mốc NAV quy ước cuối kỳ.</p>
      <label><input type="checkbox" required> Đã đối chiếu sổ, số dư, phí/thuế và giá với sao kê; đồng ý khóa kỳ</label><button>Chốt NAV & khóa kỳ</button></form></details>
      <details><summary>Nhập/xuất và sửa sai trước khi chốt</summary><div class="fundActions"><button type="button" data-template>Tải mẫu JSON</button><button type="button" data-export>Xuất sổ + đối soát</button><label>Nhập giao dịch JSON<input data-import type="file" accept=".json,application/json"></label></div>
      <p>Nhập nguyên lô hoặc không nhập nếu có lỗi. Cùng mã đối chiếu và cùng nội dung không tạo bản trùng. Hủy chỉ áp dụng kỳ chưa khóa, không xóa lịch sử; sau đó nhập lại với mã đối chiếu mới.</p>
      <form data-void><div class="fundForm">${input('entry_id','ID giao dịch cần hủy','number','','required min="1"')}${input('reason','Lý do hủy','text','','required minlength="5"')}</div><button>Hủy có lưu vết</button></form></details>
      <p data-status role="status" aria-live="polite"></p>
      <details open><summary>Sổ giao dịch — toàn lịch sử</summary><div data-entries></div></details>
      <details><summary>Lịch sử giá đã kiểm chứng — toàn lịch sử</summary><div data-prices></div></details>
      <details open><summary>Chất lượng giá từ bộ nhớ nhà cung cấp — hiện tại</summary><p>Kiểm tra bản lưu trên tín hiệu mới nhất mỗi mã, không gọi nhà cung cấp trực tiếp. Tuổi giá tính theo ngày lịch; trên 4 ngày hoặc có nến lỗi cần rà soát. Giá này không tự đi vào NAV. Mã chỉ có trong sổ, chưa có webhook, sẽ không có dòng bộ nhớ giá.</p><div data-health></div></details>
      <details open><summary>Đối soát webhook — 500 tín hiệu gần nhất</summary><p>Liên kết không chứng minh khớp đủ: webhook không có khối lượng thực khớp. Chênh lệch giá so với đóng cửa cùng ngày nhận chỉ là dấu hiệu rà soát, không khẳng định giá webhook sai. |Lệch| ≥ 5% được đánh dấu. Giá webhook cổ phiếu từ nghìn VND được đổi sang VND.</p><div data-audit></div></details>
      <details><summary>Webhook bị phân lỗi — 100 bản ghi gần nhất</summary><div data-invalid></div></details>`;
    panel.querySelector('[data-reload]').onclick=()=>action(async()=>{},'Đã tải dữ liệu.');
    panel.querySelector('[data-entry]').onsubmit=e=>{
      e.preventDefault();const d=Object.fromEntries(new FormData(e.target));
      for(const k of ['quantity','price','fees','amount'])d[k]=Number(d[k]);d.signal_id=d.signal_id?Number(d.signal_id):null;
      action(()=>api('/entries',{entries:[d]}),'Đã ghi nhận (mã trùng giống nội dung sẽ được bỏ qua).');
    };
    panel.querySelector('[data-price]').onsubmit=e=>{e.preventDefault();const d=Object.fromEntries(new FormData(e.target));d.price=Number(d.price);d.confirmed=true;action(()=>api('/prices',d),'Đã lưu giá đã kiểm chứng.');};
    panel.querySelector('[data-close]').onsubmit=e=>{e.preventDefault();const d=Object.fromEntries(new FormData(e.target));d.benchmark_value=d.benchmark_value?Number(d.benchmark_value):null;action(()=>api('/close',d),'Đã chốt NAV và khóa kỳ. Các bảng hiệu suất/rủi ro cập nhật ở lần làm mới kế tiếp.');};
    panel.querySelector('[data-void]').onsubmit=e=>{e.preventDefault();const d=Object.fromEntries(new FormData(e.target));action(()=>api(`/entries/${Number(d.entry_id)}/void`,{reason:d.reason}),'Đã hủy có lưu lịch sử.');};
    panel.querySelector('[data-template]').onclick=()=>download('ledger-template.json',{entries:[{reference:'EXAMPLE-deposit-001',trade_date:day,kind:'deposit',amount:100000000},{reference:'EXAMPLE-buy-001',trade_date:day,kind:'buy',ticker:'MBB',sleeve:'RF',sector:'banking',quantity:200,price:25000,fees:7500,note:'Ví dụ: thay toàn bộ bằng số liệu sao kê'}]});
    panel.querySelector('[data-export]').onclick=()=>{if(latest)download('ledger-audit.json',latest);};
    panel.querySelector('[data-import]').onchange=e=>{const f=e.target.files[0];if(!f)return;action(async()=>{if(f.size>5000000)throw new Error('Tệp tối đa 5 MB');const d=JSON.parse(await f.text());return api('/entries',{entries:Array.isArray(d)?d:d.entries});},'Đã nhập giao dịch.');e.target.value='';};
  }
  function render(d) {
    latest=d;
    const cards=[['NAV dự thảo VND',d.nav],['Tiền mặt kế toán VND',d.cash],['Lãi/lỗ đã chốt VND',d.realized_pnl],['Lãi/lỗ chưa chốt VND',d.unrealized_pnl],['Cổ tức ròng VND',d.dividend_income],['Chi phí khác VND',d.expenses],['Tổng lãi/lỗ VND',d.total_pnl]];
    panel.querySelector('[data-summary]').innerHTML=`<p>Định giá ${esc(d.trade_date)} • Đã khóa đến: ${esc(d.locked_through||'chưa khóa')} • ${d.signal_count} tín hiệu lưu, ${d.invalid_count} bản ghi phân lỗi • ${d.unlinked_trades} giao dịch không gắn webhook.</p><div class="fundKpis">${cards.map(([k,v])=>`<article><span>${esc(k)}</span><strong>${num(v)}</strong></article>`).join('')}</div>${d.missing_prices.length?`<p class="fundNotice">Chưa thể tính/chốt NAV. Thiếu giá đã kiểm chứng cùng ngày: ${esc(d.missing_prices.join(', '))}</p>`:''}${table(['Mã','Nhóm','Số lượng','Giá vốn BQ gồm phí','Giá VND','Giá trị VND','P/L chưa chốt'],d.holdings.map(h=>[h.ticker,h.sleeve,num(h.quantity),num(h.cost/h.quantity),num(h.price),num(h.market_value),num(h.unrealized_pnl)]))}`;
    panel.querySelector('[data-entries]').innerHTML=table(['ID','Ngày','Mã đối chiếu','Loại','Mã','Nhóm','SL','Giá VND','Phí/thuế','Dòng tiền','Webhook','Người nhập','Trạng thái'],[...d.entries].reverse().map(e=>[e.id,e.trade_date,e.reference,labels[e.kind],e.ticker,e.sleeve,num(e.quantity),num(e.price),num(e.fees),num(e.amount),e.signal_id,e.created_by,e.void_reason?`Hủy: ${e.void_reason}`:'Có hiệu lực']));
    panel.querySelector('[data-prices]').innerHTML=table(['ID','Ngày','Mã','Giá VND','Nguồn','Tham chiếu','Người duyệt','Lúc lưu'],[...d.prices].reverse().map(p=>[p.id,p.trade_date,p.ticker,num(p.price),p.source,p.evidence,p.created_by,p.created_at]));
    panel.querySelector('[data-health]').innerHTML=table(['Mã','Trạng thái nguồn','Ngày giá mới nhất','Tuổi (ngày)','Giá lưu VND','Nến lỗi','Lệch giá kiểm chứng %','Đánh giá'],d.price_health.map(p=>[p.ticker,p.status,p.latest_date,p.age_days,num(p.cached_price_vnd),p.invalid_bars,num(p.deviation_pct),!p.latest_date?'Thiếu giá':p.age_days>4||p.invalid_bars||p.status!=='ok'||Math.abs(p.deviation_pct)>=5?'Cần rà soát':'Chưa thấy bất thường cấu trúc']));
    panel.querySelector('[data-audit]').innerHTML=table(['ID','Nhận lúc','Mã','Chiến lược','Tín hiệu','Trạng thái','Lần khớp','SL khớp','Giá webhook VND','Giá kiểm chứng VND','Lệch %'],d.webhooks.map(w=>[w.id,w.received_at,w.ticker,w.strategy,w.action,w.status,w.fill_count,num(w.filled_quantity),num(w.signal_price_vnd),num(w.reference_price),`${num(w.deviation_pct)}${Math.abs(w.deviation_pct)>=5?' ⚠':''}`]));
    panel.querySelector('[data-invalid]').innerHTML=table(['ID','Lúc nhận','Mã','Tín hiệu','Phân lỗi'],d.invalid.map(w=>[w.id,w.received_at,w.ticker,w.action,w.reason]));
  }
  async function load() {
    const current=++version;
    const day=panel.querySelector('[name="valuation_date"]').value;
    const d=await api(`?trade_date=${encodeURIComponent(day)}`);
    if(current===version)render(d);
  }
  function clear(){version++;identity=null;latest=null;panel.replaceChildren();panel.hidden=true;}
  async function refresh(user){
    if(user?.role!=='admin'){clear();return;}
    if(identity!==user.id){clear();identity=user.id;panel.hidden=false;mount();}
    const currentIdentity=identity;
    try{await load();}catch(e){if(identity===currentIdentity)panel.querySelector('[data-status]').textContent=e.message;}
  }
  window.TradeLedger={clear,refresh};
})();

File: app/services/trade_ledger.py
"""Confirmed executions, independent of alerts. All monetary inputs are VND."""
from __future__ import annotations

import json
import math
from datetime import date, datetime
from decimal import Decimal
from typing import Literal
from zoneinfo import ZoneInfo

from pydantic import Field, field_validator, model_validator

from app.database import utc_now_iso
from app.services.fund_analytics import AnalyticsModel, NavSnapshot


def today() -> date:
    return datetime.now(ZoneInfo("Asia/Ho_Chi_Minh")).date()


class Dated(AnalyticsModel):
    trade_date: date

    @field_validator("trade_date")
    @classmethod
    def not_future(cls, value):
        if value > today():
            raise ValueError("Không ghi nhận ngày tương lai")
        return value


class Entry(Dated):
    reference: str = Field(min_length=1, max_length=100)
    kind: Literal["deposit", "withdraw", "buy", "sell", "dividend", "expense"]
    ticker: str = Field(default="", max_length=20, pattern=r"^[A-Za-z0-9._-]*$")
    sleeve: Literal["RF", "EMA", "OTHER"] = "OTHER"
    sector: str = Field(default="unknown", min_length=1, max_length=100)
    quantity: float = Field(default=0, ge=0, le=1e12)
    price: float = Field(default=0, ge=0, le=1e12)
    amount: float = Field(default=0, ge=0, le=1e18)
    fees: float = Field(default=0, ge=0, le=1e15)
    signal_id: int | None = Field(default=None, ge=1)
    note: str = Field(default="", max_length=1000)

    @model_validator(mode="after")
    def shape(self):
        self.ticker = self.ticker.upper()
        self.sector = self.sector.casefold()
        if self.kind in {"buy", "sell"}:
            if not self.ticker or not self.quantity or not self.price or self.amount:
                raise ValueError("Mua/bán cần mã, số lượng, giá VND; amount phải bằng 0")
        elif not self.amount or self.quantity or self.price or self.fees or self.signal_id:
            raise ValueError("Dòng tiền cần amount > 0; quantity, price, fees bằng 0 và không gắn webhook")
        if self.kind == "dividend" and not self.ticker:
            raise ValueError("Cổ tức cần mã cổ phiếu; nhập số tiền thực nhận sau thuế")
        return self


class EntryBatch(AnalyticsModel):
    entries: list[Entry] = Field(min_length=1, max_length=2000)


class VerifiedPrice(Dated):
    ticker: str = Field(min_length=1, max_length=20, pattern=r"^[A-Za-z0-9._-]+$")
    price: float = Field(gt=0, le=1e12)
    source: str = Field(min_length=3, max_length=200)
    evidence: str = Field(min_length=3, max_length=1000)
    confirmed: Literal[True]

    @field_validator("ticker")
    @classmethod
    def uppercase(cls, value):
        return value.upper()


class VoidEntry(AnalyticsModel):
    reason: str = Field(min_length=5, max_length=1000)


class CloseRequest(Dated):
    benchmark_name: str = Field(default="VNINDEX (price)", min_length=1, max_length=100)
    benchmark_value: float | None = Field(default=None, gt=0, le=1e12)


def entries(conn):
    return [{**json.loads(r["entry_json"]), **{k: r[k] for k in (
        "id", "created_by", "created_at", "void_reason", "voided_by", "voided_at")}}
        for r in conn.execute("SELECT * FROM ledger_entries ORDER BY id")]


def prices(conn):
    return [{**json.loads(r["price_json"]), "id": r["id"], "created_by": r["created_by"],
             "created_at": r["created_at"]} for r in conn.execute("SELECT * FROM ledger_prices ORDER BY id")]


def locked(conn, day):
    last = conn.execute("SELECT MAX(trade_date) FROM ledger_closes").fetchone()[0]
    if last and day <= last:
        raise ValueError(f"Sổ đã khóa đến {last}; không sửa dữ liệu lịch sử đã chốt NAV")


def replay(rows, day):
    """Weighted average cost includes buy fees; sell fees reduce realized P/L."""
    cash = realized = flows = income = expenses = Decimal(0)
    holdings = {}
    D = lambda x: Decimal(str(x))
    for row in sorted(rows, key=lambda r: (r["trade_date"], r.get("id", 0))):
        if row.get("void_reason") or row["trade_date"] > day:
            continue
        kind = row["kind"]
        amount, qty, price, fees = (D(row[k]) for k in ("amount", "quantity", "price", "fees"))
        if kind == "deposit":
            cash += amount
            flows += amount
        elif kind == "withdraw":
            cash -= amount
            flows -= amount
        elif kind == "dividend":
            cash += amount
            income += amount
        elif kind == "expense":
            cash -= amount
            expenses += amount
        else:
            key = (row["ticker"], row["sleeve"])
            h = holdings.setdefault(key, dict(ticker=key[0], sleeve=key[1], sector=row["sector"],
                                              quantity=Decimal(0), cost=Decimal(0)))
            if h["quantity"] and h["sector"] != row["sector"]:
                raise ValueError(f"Ngành không nhất quán: {row['ticker']}")
            h["sector"] = row["sector"]
            if kind == "buy":
                cash -= qty * price + fees
                h["quantity"] += qty
                h["cost"] += qty * price + fees
            else:
                if qty > h["quantity"]:
                    raise ValueError(f"Bán vượt số lượng: {row['reference']}")
                cost = h["cost"] * qty / h["quantity"]
                h["quantity"] -= qty
                h["cost"] -= cost
                cash += qty * price - fees
                realized += qty * price - fees - cost
        if cash < 0:
            raise ValueError(f"Tiền mặt âm sau {row['reference']}; kiểm tra thứ tự giao dịch/nạp tiền")
    sectors = {}
    for h in holdings.values():
        if h["quantity"]:
            if h["ticker"] in sectors and sectors[h["ticker"]] != h["sector"]:
                raise ValueError(f"Một mã phải cùng ngành ở các nhóm: {h['ticker']}")
            sectors[h["ticker"]] = h["sector"]
    return dict(cash=float(cash), realized_pnl=float(realized), external_flows=float(flows),
                dividend_income=float(income), expenses=float(expenses), holdings=[
                    {**h, "quantity": float(h["quantity"]), "cost": float(h["cost"])}
                    for h in holdings.values() if h["quantity"]])


def add_entries(store, batch, user_id):
    with store.connect() as conn:
        conn.execute("BEGIN IMMEDIATE")
        count = 0
        for item in batch.entries:
            row = item.model_dump(mode="json")
            encoded = json.dumps(row, ensure_ascii=False, sort_keys=True)
            existing = conn.execute("SELECT entry_json, void_reason FROM ledger_entries WHERE reference=?", (item.reference,)).fetchone()
            if existing:
                if existing["entry_json"] == encoded and not existing["void_reason"]:
                    continue
                raise ValueError(f"Mã đối chiếu đã tồn tại: {item.reference}")
            locked(conn, row["trade_date"])
            if item.signal_id:
                signal = conn.execute("SELECT * FROM signals WHERE id=?", (item.signal_id,)).fetchone()
                expected = {"buy": {"buy", "confirm_buy"}, "sell": {"sell", "confirm_sell"}}[item.kind]
                if not signal or signal["ticker"] != item.ticker or signal["action"] not in expected:
                    raise ValueError("Webhook liên kết không tồn tại hoặc sai mã/chiều giao dịch")
                received_day = datetime.fromisoformat(signal["received_at"]).astimezone(ZoneInfo("Asia/Ho_Chi_Minh")).date()
                if item.trade_date < received_day:
                    raise ValueError("Ngày khớp không được trước ngày nhận webhook liên kết")
            conn.execute("INSERT INTO ledger_entries(reference,entry_json,created_by,created_at) VALUES(?,?,?,?)",
                         (item.reference, encoded, user_id, utc_now_iso()))
            count += 1
        replay(entries(conn), today().isoformat())
    return {"inserted": count}


def void_entry(store, entry_id, reason, user_id):
    with store.connect() as conn:
        conn.execute("BEGIN IMMEDIATE")
        row = conn.execute("SELECT * FROM ledger_entries WHERE id=?", (entry_id,)).fetchone()
        if not row or row["void_reason"]:
            raise ValueError("Giao dịch không tồn tại hoặc đã hủy")
        locked(conn, json.loads(row["entry_json"])["trade_date"])
        conn.execute("UPDATE ledger_entries SET void_reason=?,voided_by=?,voided_at=? WHERE id=?",
                     (reason, user_id, utc_now_iso(), entry_id))
        replay(entries(conn), today().isoformat())
    return {"status": "voided"}


def add_price(store, item, user_id):
    with store.connect() as conn:
        conn.execute("BEGIN IMMEDIATE")
        locked(conn, item.trade_date.isoformat())
        conn.execute("INSERT INTO ledger_prices(price_json,created_by,created_at) VALUES(?,?,?)",
                     (item.model_dump_json(), user_id, utc_now_iso()))
    return {"status": "verified"}


def valuation(conn, day):
    rows = entries(conn)
    result = replay(rows, day)
    quotes = {p["ticker"]: p for p in prices(conn) if p["trade_date"] == day}
    missing = []
    market = unrealized = 0
    for h in result["holdings"]:
        p = quotes.get(h["ticker"])
        h.update(price=p["price"] if p else None, price_source=p["source"] if p else None)
        h["market_value"] = h["quantity"] * h["price"] if p else None
        h["unrealized_pnl"] = h["market_value"] - h["cost"] if p else None
        if p:
            market += h["market_value"]
            unrealized += h["unrealized_pnl"]
        else:
            missing.append(h["ticker"])
    result.update(trade_date=day, missing_prices=sorted(set(missing)),
                  nav=None if missing else result["cash"] + market,
                  unrealized_pnl=None if missing else unrealized,
                  external_flow=sum((r["amount"] if r["kind"] == "deposit" else -r["amount"])
                      for r in rows if not r["void_reason"] and r["trade_date"] == day and r["kind"] in {"deposit", "withdraw"}))
    result["total_pnl"] = None if missing else result["nav"] - result["external_flows"]
    return result


def close_nav(store, item, user_id):
    day = item.trade_date.isoformat()
    with store.connect() as conn:
        conn.execute("BEGIN IMMEDIATE")
        locked(conn, day)
        v = valuation(conn, day)
        if v["missing_prices"]:
            raise ValueError("Thiếu giá cùng ngày đã kiểm chứng: " + ", ".join(v["missing_prices"]))
        if conn.execute("SELECT 1 FROM nav_snapshots WHERE trade_date>=?", (day,)).fetchone():
            raise ValueError("Chỉ nối NAV mới sau ngày NAV cuối cùng; không ghi đè lịch sử")
        previous = conn.execute("SELECT MAX(trade_date) FROM nav_snapshots").fetchone()[0]
        if previous and not conn.execute("SELECT 1 FROM ledger_closes LIMIT 1").fetchone():
            baseline = valuation(conn, previous)
            old = json.loads(conn.execute("SELECT snapshot_json FROM nav_snapshots WHERE trade_date=?", (previous,)).fetchone()[0])
            old_qty = {(h["ticker"], h["sleeve"]): h["quantity"] for h in old["holdings"] if h["quantity"]}
            new_qty = {(h["ticker"], h["sleeve"]): h["quantity"] for h in baseline["holdings"]}
            if (baseline["nav"] is None or abs(baseline["nav"] - old["nav"]) > 1
                    or abs(baseline["cash"] - old["cash"]) > 1 or old_qty != new_qty
                    or old.get("receivables", 0) or old.get("liabilities", 0)):
                raise ValueError("Sổ và giá tại ngày NAV cũ chưa khớp số dư ban đầu; đối chiếu trước khi nối chuỗi NAV")
        # Aggregate all flows since previous observation, including missing NAV days.
        flow = sum((r["amount"] if r["kind"] == "deposit" else -r["amount"])
                   for r in entries(conn) if not r["void_reason"] and r["kind"] in {"deposit", "withdraw"}
                   and (not previous or previous < r["trade_date"]) and r["trade_date"] <= day)
        snapshot = NavSnapshot(trade_date=day, nav=v["nav"], cash=v["cash"], external_flow=flow if previous else 0,
            holdings=[{k: h[k] for k in ("ticker", "sleeve", "sector", "quantity", "price")} for h in v["holdings"]],
            benchmark_name=item.benchmark_name, benchmark_value=item.benchmark_value,
            note="Sổ giao dịch đã chốt; giá VND được quản trị viên kiểm chứng. Dòng tiền quy ước cuối kỳ.").model_dump(mode="json")
        encoded, now = json.dumps(snapshot, ensure_ascii=False), utc_now_iso()
        conn.execute("INSERT INTO nav_snapshots VALUES(?,?,?,?)", (day, encoded, user_id, now))
        conn.execute("INSERT INTO nav_snapshot_revisions(trade_date,snapshot_json,updated_by,updated_at) VALUES(?,?,?,?)", (day, encoded, user_id, now))
        conn.execute("INSERT INTO ledger_closes VALUES(?,?,?,?)", (day, encoded, user_id, now))
    return {"snapshot": snapshot}


def report(store, day):
    with store.connect() as conn:
        result = valuation(conn, day)
        result["entries"] = entries(conn)
        result["prices"] = prices(conn)
        result["locked_through"] = conn.execute("SELECT MAX(trade_date) FROM ledger_closes").fetchone()[0]
        signals = conn.execute("SELECT id,ticker,action,strategy,price,received_at FROM signals ORDER BY id DESC LIMIT 500").fetchall()
        result["signal_count"] = conn.execute("SELECT COUNT(*) FROM signals").fetchone()[0]
        result["invalid_count"] = conn.execute("SELECT COUNT(*) FROM invalid_signals").fetchone()[0]
        result["invalid"] = [dict(r) for r in conn.execute("SELECT id,ticker,action,reason,received_at FROM invalid_signals ORDER BY id DESC LIMIT 100")]
        cached = conn.execute("SELECT ticker,enrichment_json FROM signals WHERE id IN (SELECT MAX(id) FROM signals GROUP BY ticker)").fetchall()
    active = [e for e in result["entries"] if not e["void_reason"]]
    quotes = {(p["ticker"], p["trade_date"]): p for p in result["prices"]}
    audit = []
    for signal in signals:
        r = dict(signal)
        fills = [e for e in active if e["signal_id"] == r["id"]]
        stamp = datetime.fromisoformat(r["received_at"]).astimezone(ZoneInfo("Asia/Ho_Chi_Minh")).date().isoformat()
        p = quotes.get((r["ticker"], stamp))
        # Existing signal.price is normalized to thousands of VND for Vietnamese equities.
        r["signal_price_vnd"] = r.pop("price") * 1000 if r["price"] else None
        r["reference_price"] = p["price"] if p else None
        r["deviation_pct"] = (r["signal_price_vnd"] / p["price"] - 1) * 100 if p and r["signal_price_vnd"] else None
        r["fill_count"] = len(fills)
        r["filled_quantity"] = sum(e["quantity"] for e in fills)
        r["status"] = "Đã liên kết khớp lệnh" if fills else "Chưa liên kết khớp lệnh"
        audit.append(r)
    result["webhooks"] = audit
    health = []
    for row in cached:
        enrichment = json.loads(row["enrichment_json"])
        history = enrichment.get("history") or []
        if not isinstance(history, list):
            history = [None]
        usable = []
        invalid_bars = 0
        for bar in history:
            try:
                raw_date = str(bar.get("time", ""))
                bar_day = date.fromisoformat(raw_date[:10]).isoformat()
                close = float(bar["close"])
                if not math.isfinite(close) or not 0 < close <= 1e9 or bar_day > today().isoformat():
                    raise ValueError("invalid bar")
                usable.append((bar_day, close))
            except (ValueError, TypeError, KeyError, AttributeError):
                invalid_bars += 1
        last = max(usable, default=None)
        quote = quotes.get((row["ticker"], last[0])) if last else None
        price_vnd = last[1] * 1000 if last else None
        health.append(dict(ticker=row["ticker"], status=enrichment.get("status", "unknown"),
                           refreshed_at=enrichment.get("refreshed_at"), latest_date=last[0] if last else None,
                           cached_price_vnd=price_vnd, age_days=(today()-date.fromisoformat(last[0])).days if last else None,
                           invalid_bars=invalid_bars,
                           deviation_pct=(price_vnd/quote["price"]-1)*100 if quote else None))
    result["price_health"] = health
    result["unlinked_trades"] = sum(e["kind"] in {"buy", "sell"} and not e["signal_id"] for e in active)
    return result
